pkg/abcupdater/version.go

```diff
@@ -93,84 +92,27 @@ To disable notifications for this new version, set {{.OptOutEnvVar}}="{{.Current
-	err := localstore.LoadJSONFile(path, &cached)
-	if err != nil {
-		return nil, fmt.Errorf("could not load cached data: %w", err)
-	}
-	return &cached, nil
-}
-
-func (c *CheckVersionParams) setLocalCachedData(data *LocalVersionData) error {
-	var path string
-	if c.CacheFileOverride != "" {
-		path = c.CacheFileOverride
-	} else {
-		dir, err := localstore.DefaultDir(c.AppID)
-		if err != nil {
-			return fmt.Errorf("could not calculate cache path: %w", err)
-		}
-		path = filepath.Join(dir, localVersionFileName)
-	}
-	if err := localstore.StoreJSONFile(path, data); err != nil {
-		return fmt.Errorf("could not cache version: %w", err)
-	}
-	return nil
-}
-
 // CheckAppVersion calls CheckAppVersionSync in a go routine. It returns a closure
 // to be run after program logic which will block until a response is returned
-// or a timeout is hit. Results will be printed to stderr.
-func CheckAppVersion(ctx context.Context, params *CheckVersionParams) func() {
-	return asyncFunctionCall(ctx, func() (string, error) {
-		return CheckAppVersionSync(ctx, params)
-	},
-		os.Stderr)
-}
-
-// asyncFunctionCall handles the async part of CheckAppVersion, but accepts
-// a function and writer as an argument to allow for testability.
-func asyncFunctionCall(ctx context.Context, funcToCall func() (string, error), out io.Writer) func() {
-	updatesCh := make(chan string, 1)
-
-	go func() {
-		defer close(updatesCh)
-		message, err := funcToCall()
-		if err != nil {
-			logging.FromContext(ctx).WarnContext(ctx, "failed to check for new versions",
-				"error", err)
-		}
-		updatesCh <- message
-	}()
-
-	return func() {
-		select {
-		case <-ctx.Done():
-			// Context was cancelled
-		case msg := <-updatesCh:
-			if len(msg) > 0 {
-				fmt.Fprintf(out, "%s\n", msg)
-			}
-		case <-time.After(time.Second):
-			// Give up after some time. This is how long to wait after termination has
-			// finished. The command has most likely be running for a few seconds, and
-			// we don't want to block the control flow for an exceedingly long time.
-			//
-			// This technically leaks both the timer and the channel, but we're about
-			// to terminate, so the memory will free anyway.
-		}
+// or a timeout is hit. If there is any output, out() will be run during the
+// returned closure, for the consumer to print to user.
+// Example out(): `func(s string) {fmt.Fprintf(os.Stderr, "%s\n", s)}`
+func CheckAppVersion(ctx context.Context, params *CheckVersionParams, out func(string)) (func(), error) {
+	lookuper := params.Lookuper
+	if lookuper == nil {
+		lookuper = envconfig.OsLookuper()
+	}
+	var c config
+	if err := envconfig.ProcessWith(ctx, &envconfig.Config{
+		Target:   &c,
+		Lookuper: lookuper,
+	}); err != nil {
+		return nil, fmt.Errorf("failed to process envconfig: %w", err)
 	}
+	// Add 100ms to the timeout configured to HTTP call for time to wait on goroutine.
+	return asyncFunctionCall(ctx, c.RequestTimeout+100*time.Millisecond, func() (string, error) {
```

Comment:
What happens if ctx already has a deadline on it? Maybe we should document which timeout wins.

Reply:
This is the same behavior of any derrived context, https://go.dev/blog/context, the earliest cancellation wins.

I'm not sure it makes sense to document idomatic go

Reply:
Maybe I'm missing your point, but I don't think it's idiomatic for a function to accept a timeout and also a context (why not just pass a context alone, which already provides the timeout functionality?). I could be wrong though, my understanding of conventions has been warped by being in the google bubble.

Reply:
Yeah, I see now. You are right, I think I'll change to only accept a context

Reply:
Though that does make it less turnkey for a user....

Hard choice